File: provider/src/main/java/com/leo/sp/provider/SPContentProvider.java
package com.leo.sp.provider;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Map;
import java.util.Set;

/**
 * 依赖share存储的ContentProvider
 */
public class SPContentProvider extends ContentProvider {

    public static final int URI_TYPE_INDEX = 1;

    public static final int URI_TYPE_VALUE_KEY_INDEX = 2;

    @Override
    public boolean onCreate() {
        SpResolver.getInstance().init(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection,
                        @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        String[] path = uri.getPath().split(SpContants.SEPARATOR);
        String type = path[URI_TYPE_INDEX];
        if (type.equals(SpContants.TYPE_GET_ALL)) {
            Map<String, ?> allSp = SPImpl.getAll(getContext());
            if (allSp != null) {
                MatrixCursor cursor = new MatrixCursor(new String[]{SpContants.CURSOR_COLUMN_NAME,
                        SpContants.CURSOR_COLUMN_TYPE, SpContants.CURSOR_COLUMN_VALUE});
                Object[] rows = new Object[3];
                Set<String> ketSet = allSp.keySet();
                for (String key : ketSet) {
                    rows[0] = key;
                    rows[2] = allSp.get(key);

                    if (rows[2] instanceof Boolean) {
                        rows[1] = SpContants.TYPE_BOOLEAN;
                    } else if (rows[2] instanceof String) {
                        rows[1] = SpContants.TYPE_STRING;
                    } else if (rows[2] instanceof Integer) {
                        rows[1] = SpContants.TYPE_INT;
                    } else if (rows[2] instanceof Long) {
                        rows[1] = SpContants.TYPE_LONG;
                    } else if (rows[2] instanceof Float) {
                        rows[1] = SpContants.TYPE_FLOAT;
                    }
                    cursor.addRow(rows);
                }
                return cursor;
            }
        }
        return null;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        String[] path = uri.getPath().split(SpContants.SEPARATOR);
        String type = path[URI_TYPE_INDEX];
        String key = path[URI_TYPE_VALUE_KEY_INDEX];
        if (type.equals(SpContants.TYPE_CONTAIN)) {
            return SPImpl.contains(getContext(), key) + "";
        }
        return "" + SPImpl.get(getContext(), key, type);
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        if (null == values) {
            return null;
        }
        String[] path = uri.getPath().split(SpContants.SEPARATOR);
        String type = path[1];
        String key = path[2];
        Object obj = values.get(SpContants.VALUE);
        if (obj != null) {
            SPImpl.save(getContext(), key, obj);
        }
        getContext().getContentResolver().notifyChange(uri, null);
        return null;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        String[] path = uri.getPath().split(SpContants.SEPARATOR);
        String type = path[1];
        if (type.equals(SpContants.TYPE_CLEAN)) {
            SPImpl.clear(getContext());
            return 0;
        }
        String key = path[2];
        if (SPImpl.contains(getContext(), key)) {
            SPImpl.remove(getContext(), key);
        }
        getContext().getContentResolver().notifyChange(uri, null);
        return 0;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection,
                      @Nullable String[] selectionArgs) {
        insert(uri, values);
        return 0;
    }
}

File: provider/src/main/java/com/leo/sp/provider/SPContentProvider.kt
package com.leo.sp.provider

import android.content.ContentProvider
import android.content.ContentValues
import android.content.Context
import android.database.ContentObserver
import android.database.Cursor
import android.net.Uri
import android.util.Log
import androidx.annotation.Keep
import com.leo.sp.provider.datastore.SpDataStoreImpl
import kotlin.reflect.KClass

/**
 * 依赖share存储的ContentProvider
 */
@Keep
class SPContentProvider : ContentProvider() {
    private lateinit var mCtx: Context
    override fun onCreate(): Boolean {
        mCtx = context!!
        SpContants.initAuthority(mCtx)
        SpResolver.getInstance().init(mCtx)
        return true
    }

    override fun query(
        uri: Uri, projection: Array<String>?, selection: String?,
        selectionArgs: Array<String>?, sortOrder: String?
    ): Cursor? {
        return null
    }

    override fun getType(uri: Uri): String? {
        val path =
            getSafePath(uri).split(SpContants.SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()
        val type = path[URI_TYPE_INDEX]
        val key = path[URI_TYPE_VALUE_KEY_INDEX]
        return if (type == SpContants.TYPE_CONTAIN) {
            SpDataStoreImpl.containsSync(mCtx, key).toString()
        } else {
            val value = SpDataStoreImpl.getSync(mCtx, key, getTypeClass(type))
            return value.toString()
        }
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        if (null == values) {
            return null
        }
        val path =
            getSafePath(uri).split(SpContants.SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()
        val type = path[1]
        val key = path[2]
        val obj = values[SpContants.VALUE]
        if (obj != null) {
            SpDataStoreImpl.put(mCtx, key, obj, getTypeClass(type))
        }
        safeNotifyChange(uri, null)
        return null
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val path =
            getSafePath(uri).split(SpContants.SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()
        val type = path[1]
        if (type == SpContants.TYPE_CLEAN) {
            SpDataStoreImpl.clear(mCtx)
            return 0
        }
        val key = path[2]
        SpDataStoreImpl.remove(mCtx, key, getTypeClass(type))
        safeNotifyChange(uri, null)
        return 0
    }

    override fun update(
        uri: Uri, values: ContentValues?, selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        insert(uri, values)
        return 0
    }

    private fun getTypeClass(type: String): KClass<out Any> {
        return when (type) {
            SpContants.TYPE_STRING -> String::class
            SpContants.TYPE_INT -> Int::class
            SpContants.TYPE_LONG -> Long::class
            SpContants.TYPE_FLOAT -> Float::class
            SpContants.TYPE_DOUBLE -> Double::class
            SpContants.TYPE_BOOLEAN -> Boolean::class
            else -> throw RuntimeException("getTypeClass: not support type -> $type")
        }
    }

    private fun getSafePath(uri: Uri): String {
        return if (uri.path == null) "" else uri.path!!
    }

    private fun safeNotifyChange(uri: Uri, observer: ContentObserver?) {
        val contentResolver = mCtx.contentResolver
        if (contentResolver == null) {
            Log.e(TAG, "safeNotifyChange: contentResolver is null.")
            return
        }
        contentResolver.notifyChange(uri, observer)
    }

    companion object {
        private const val TAG = "SPContentProvider"
        const val URI_TYPE_INDEX = 1
        const val URI_TYPE_VALUE_KEY_INDEX = 2
    }
}
